### ISSUE TITLE
make_wrn_inception

### ISSUE THREAD

**Issue by congmonkey** [bob48523/modified_wrn#5]
from __future__ import print_function

import math
from caffe import layers as L, params as P, to_proto
from caffe.proto import caffe_pb2
import caffe

def bn_relu_conv(bottom, ks, nout, stride, pad, dropout):
    batch_norm = L.BatchNorm(bottom, in_place=False, param=[dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0)])
    scale = L.Scale(batch_norm, bias_term=True, in_place=True, filler=dict(value=1), bias_filler=dict(value=0))
    relu = L.ReLU(scale, in_place=True)
    conv = L.Convolution(relu, kernel_size=ks, stride=stride, 
                    num_output=nout, pad=pad, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
    if dropout>0:
        conv = L.Dropout(conv, dropout_ratio=dropout)
    return conv

def inception(bottom, stride, nout):
      
    outplane1 = nout // 2
    outplane2 = nout - outplane1 
    
    conv1 = bn_relu_conv(bottom, ks=1, nout=64, stride=stride, pad=0, dropout=0)
    conv1 = bn_relu_conv(conv1, ks=3, nout=outplane1, stride=1, pad=0, dropout=0)
    branch1 = bn_relu_conv(conv1, ks=3, nout=outplane1, stride=1, pad=0, dropout=0)

    conv2 = bn_relu_conv(bottom, ks=1, nout=48, stride=stride, pad=0, dropout=0)
    branch2 = bn_relu_conv(conv2, ks=3, nout=outplane2, stride=1, pad=0, dropout=0)
    
    concate = L.Concat(branch1, branch2, axis=1)
    
    return concate
    
def shortcut(bottom, nout, stride):
    conv = L.Convolution(bottom, kernel_size=1, stride=stride, 
                    num_output=nout, pad=0, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
   
    return conv

def add_layer(bottom, num_filter, stride, first):
    
    if stride!=1 or first == 1:
        x = shortcut(bottom, nout=num_filter, stride=stride)
    else:
        x = bottom
    
    conv = inception(bottom, stride=stride, nout=num_filter)
       
    out = L.Eltwise(x, conv)
    return out

def transition(bottom, num_filter, dropout):
    conv = bn_relu_conv(bottom, ks=1, nout=num_filter, stride=1, pad=0, dropout=dropout)
    pooling = L.Pooling(conv, pool=P.Pooling.AVE, kernel_size=2, stride=2)
    return pooling

#change the line below to experiment with different setting
#depth -- must be 3n+4
#first_output -- channels before entering the first dense block, twice the growth_rate for DenseNet-BC
#growth_rate -- growth rate
#dropout -- set to 0 to disable dropout, non-zero number to set dropout rate
def wrnnet(data_file, mode='train', batch_size=64, depth=[3,4,6,3], widen_factor=4, dropout=0.3):
    data, label = L.Data(source=data_file, backend=P.Data.LMDB, batch_size=batch_size, ntop=2, 
              transform_param=dict(mean_file="/data1/zuotongchun/cifar100_mean/cifar100_train_all_mean.binaryproto"))

    nchannels = 64
    k = widen_factor
    width = [64*k, 128*k, 256*k]
    model = L.Convolution(data, kernel_size=3, stride=1, num_output=nchannels,
                        pad=1, bias_term=False, weight_filler=dict(type='msra'), bias_filler=dict(type='constant'))
    model = L.BatchNorm(model, in_place=False, param=[dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0), dict(lr_mult=0, decay_mult=0)])
    model = L.Scale(model, bias_term=True, in_place=True, filler=dict(value=1), bias_filler=dict(value=0))
    model = L.ReLU(model, in_place=True)

    # 1 x 1 convolution produce 4k feature-maps
    # nchannels_B = 4 * growth_rate
    #N = (depth-4)/3/2
    #N1 = 6
    #N2 = 12
    #N3 = 32
    #N4 = 32
    
    strides = [1] + [1]*(depth[0]-1)
    first = 1
    for stride in strides:        
        model = add_layer(model, width[0], stride, first)
        first = 0
    
    first = 1
    strides = [2] + [1]*(depth[1]-1)
    for stride in strides:        
        model = add_layer(model, width[1], stride, first)
        first = 0

    first = 1
    strides = [2] + [1]*(depth[2]-1)
    for stride in strides:        
        model = add_layer(model, width[2], stride, first)
        first = 0
        		
    model = L.Pooling(model, pool=P.Pooling.AVE, global_pooling=True)
    model = L.InnerProduct(model, num_output=10, bias_term=True, weight_filler=dict(type='xavier'), bias_filler=dict(type='constant'))
    loss = L.SoftmaxWithLoss(model, label)
    accuracy = L.Accuracy(model, label)
    return to_proto(loss, accuracy)

def make_net():

    with open('train-wrn-inception-250.prototxt', 'w') as f:
        # change the path to your data. If it's not lmdb format, also change first line of densenet() function
        print(str(wrnnet('/data1/zuotongchun/cifar100_train_all_lmdb', batch_size=64)), file=f)

    with open('test-wrn-inception-250.prototxt', 'w') as f:
        print(str(wrnnet('/data1/zuotongchun/cifar100_test_lmdb', batch_size=50)), file=f)

def make_solver():
    s = caffe_pb2.SolverParameter()
    s.random_seed = 0xCAFFE

    s.train_net = 'train-wrn-inception-250.prototxt'
    s.test_net.append('train-wrn-inception-250.prototxt')
    s.test_interval = 200
    s.test_iter.append(200)

    s.max_iter = 230000
    s.type = 'Nesterov'
    s.display = 100

    s.base_lr = 0.1
    s.momentum = 0.9
    s.weight_decay = 1e-1

    s.lr_policy='multistep'
    s.gamma = 0.1
    s.stepvalue.append(int(0.5 * s.max_iter))
    s.stepvalue.append(int(0.75 * s.max_iter))
    s.solver_mode = caffe_pb2.SolverParameter.GPU

    # add by sdh
    s.device_id = 6 # 2 or 3
    s.snapshot = 1000
    s.snapshot_prefix = 'snapshot/densenet-BC-169'

    solver_path = 'solver.prototxt'
    with open(solver_path, 'w') as f:
        f.write(str(s))
        f.write('\n')

if __name__ == '__main__':

    make_net()
    make_solver()
